<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies).

In [4]:
# текст запроса
query_3_1 = f'''
select
    count(*)
from vacancies
'''

In [5]:
# результат запроса
vac_amount = pd.read_sql_query(query_3_1, connection)
print(vac_amount)

<ipython-input-5-66fffb49e411>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_amount = pd.read_sql_query(query_3_1, connection)


   count
0  49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers).

In [6]:
# текст запроса
query_3_2 = f'''
select
    count(*)
from employers
'''

In [7]:
# результат запроса
employers_amount = pd.read_sql_query(query_3_2, connection)
print(employers_amount)

   count
0  23501


<ipython-input-7-b79b93ea9a75>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_amount = pd.read_sql_query(query_3_2, connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
select
    count(*)
from areas
'''

In [9]:
# результат запроса
areas_amount = pd.read_sql_query(query_3_3, connection)
print(areas_amount)

<ipython-input-9-5d3553aacff6>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  areas_amount = pd.read_sql_query(query_3_3, connection)


   count
0   1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
select
    count(*)
from industries
'''

In [11]:
# результат запроса
industries_amount = pd.read_sql_query(query_3_4, connection)
print(industries_amount)

<ipython-input-11-b1407f36427d>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  industries_amount = pd.read_sql_query(query_3_4, connection)


   count
0    294


***

# Выводы по предварительному анализу данных
Количество записей в наших таблицах:

1. Вакансии - 49197
2. Работодатели - 23501
3. Регионы - 1362
4. Сферы деятельности - 294

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''
select count(*) "Количество ваканский", --выводим количество вакансий
a.name "Регион" --для каждого региона
from vacancies v, areas a
--соединяем таблицы
where v.area_id = a.id --по id региона
group by a.name --группируем по названию региона
order by 1 desc --сортируем в порядке убывания количества вакансий
'''

In [13]:
# результат запроса
top_areas = pd.read_sql_query(query_4_1, connection)
print(top_areas)

<ipython-input-13-a0993ff92112>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_areas = pd.read_sql_query(query_4_1, connection)


     Количество ваканский           Регион
0                    5333           Москва
1                    2851  Санкт-Петербург
2                    2112            Минск
3                    2006      Новосибирск
4                    1892           Алматы
..                    ...              ...
764                     1       Тарко-Сале
765                     1    Новоаннинский
766                     1            Бирск
767                     1           Сасово
768                     1             Анна

[769 rows x 2 columns]


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''
select count(*) --считаем количество вакансий
from vacancies
--где хотя бы одно значение з/п не пустое
where salary_from is not null or salary_to is not null
'''

In [15]:
# результат запроса
salary_check = pd.read_sql_query(query_4_2, connection)
print(salary_check)

   count
0  24073


<ipython-input-15-ae431a31004f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_check = pd.read_sql_query(query_4_2, connection)


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''
select round(avg(salary_from)) as "from", round(avg(salary_to)) as "to"
from vacancies
'''

In [17]:
# результат запроса
salary_check_avg = pd.read_sql_query(query_4_3, connection)
print(salary_check_avg)

      from        to
0  71065.0  110537.0


<ipython-input-17-7959531696c1>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_check_avg = pd.read_sql_query(query_4_3, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''
select
    schedule "Тип рабочего графика",
    employment "Тип трудоустройства",
    count(*) --выводим количество вакансий
from vacancies
group by schedule, employment --группируем по графику и трудоустройству
order by count desc --сортируем по количеству вакансий по убыванию
'''

In [19]:
# результат запроса
top_schedules = pd.read_sql_query(query_4_4, connection)
print(top_schedules)

<ipython-input-19-35fb5e596d28>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_schedules = pd.read_sql_query(query_4_4, connection)


   Тип рабочего графика  Тип трудоустройства  count
0           Полный день     Полная занятость  35367
1      Удаленная работа     Полная занятость   7802
2         Гибкий график     Полная занятость   1593
3      Удаленная работа  Частичная занятость   1312
4        Сменный график     Полная занятость    940
5           Полный день           Стажировка    569
6        Вахтовый метод     Полная занятость    367
7           Полный день  Частичная занятость    347
8         Гибкий график  Частичная занятость    312
9           Полный день     Проектная работа    141
10     Удаленная работа     Проектная работа    133
11        Гибкий график           Стажировка    116
12       Сменный график  Частичная занятость    101
13     Удаленная работа           Стажировка     64
14        Гибкий график     Проектная работа     18
15       Сменный график           Стажировка     12
16       Вахтовый метод     Проектная работа      2
17       Сменный график     Проектная работа      1


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [20]:
# текст запроса
query_4_5 = f'''
select
    experience "Требуемый опыт работы",
    count(id)
from public.vacancies
group by experience
order by count
'''

In [21]:
# результат запроса
exp_rating = pd.read_sql_query(query_4_5, connection)
print(exp_rating)

<ipython-input-21-5efb33d5cab9>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  exp_rating = pd.read_sql_query(query_4_5, connection)


  Требуемый опыт работы  count
0           Более 6 лет   1337
1             Нет опыта   7197
2         От 3 до 6 лет  14511
3    От 1 года до 3 лет  26152


***

# Выводы по детальному анализу вакансий
Больше всего вакансий относятся к следующим регионам: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. У 24073 вакансий указан размер заработной платы. В среднем работодатели готовы платить от 71065 до 110537 рублей. Самые популярные виды графиков - "полный день/полная занятость" и "удаленная работа/полная занятость". В подавляющем большинстве вакансий требуемый опыт составляет  от 1 до 3 лет. Наименьшее число вакансий - с требуемым опытом более 6-ти лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''
select
    e.name "Работодатель", --выводим название работодателя
    count(v.id) "Кол-во вакансий" --и количество вакансий
from vacancies v, employers e
--связываем таблицы
where v.employer_id = e.id --по id работодателя
group by 1 --группируем по работодателю
order by 2 desc --сортируем по вакансиям в порядке убывания
limit 5 --обрезаем вывод данных до 5 строк, чтобы увидеть 1-ую и 5-ую позиции
'''

In [23]:
# результат запроса
top_employers = pd.read_sql_query(query_5_1, connection)
print(top_employers)

<ipython-input-23-54c3640da89d>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_employers = pd.read_sql_query(query_5_1, connection)


    Работодатель  Кол-во вакансий
0         Яндекс             1933
1     Ростелеком              491
2       Тинькофф              444
3           СБЕР              428
4  Газпром нефть              331


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = f'''
select --выводим необходимые значения
    a.name "Регион",
    count(e.id) "Кол-во работодателей",
    count(v.id) "Кол-во вакансий"
from areas a
--к таблице регионов присоединяем по id региона
left join employers e on e.area = a.id --таблицу работодателей
left join vacancies v on v.area_id = a.id --и таблицу вакансий
group by 1 --группируем по региону
order by 3, 2 desc --сортируем сначала по количеству работодателей, затем - вакансий (по убыванию)
'''

In [25]:
# результат запроса
region_emp_vac = pd.read_sql_query(query_5_2, connection)
print(region_emp_vac)

<ipython-input-25-92948973aee0>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  region_emp_vac = pd.read_sql_query(query_5_2, connection)


                  Регион  Кол-во работодателей  Кол-во вакансий
0                 Россия                   410                0
1              Казахстан                   207                0
2     Московская область                    75                0
3     Краснодарский край                    19                0
4     Ростовская область                    18                0
...                  ...                   ...              ...
1357         Новосибирск               1149438          1149438
1358              Алматы               1364132          1364132
1359               Минск               2354880          2354880
1360     Санкт-Петербург               6320667          6320667
1361              Москва              31272712         31272712

[1362 rows x 3 columns]


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = f'''
select --выводим показатели
    e.name "Работодатель",
    count(distinct v.area_id) "Кол-во регионов"
from employers e --из таблицы работодателей,
join vacancies v on e.id = v.employer_id --присоединив таблицу вакансий по id работодателя
group by 1 --группируем по работодателю
order by 2 desc --сортируем по вакансиям в порядке убывания
'''

In [27]:
# результат запроса
emp_regions_count = pd.read_sql_query(query_5_3, connection)
print(emp_regions_count)

<ipython-input-27-6d43781fe81b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_regions_count = pd.read_sql_query(query_5_3, connection)


                 Работодатель  Кол-во регионов
0                      Яндекс              181
1                  Ростелеком              152
2                  Спецремонт              116
3      Поляков Денис Иванович               88
4                   ООО ЕФИН                71
...                       ...              ...
14761                  UniSol                1
14762            UNISTORY LLC                1
14763                   UNIT6                1
14764     United Distribution                1
14765              UNITEL LLC                1

[14766 rows x 2 columns]


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [28]:
# текст запроса
query_5_4 = f'''
select
    count(e.id) --считаем количество работодателей
from employers e
--присоединяем таблицу с вакансиями
left join employers_industries ei on e.id = ei.employer_id --по id работодателя
where ei.industry_id is null --фильтруем по пустым значениям сфер деятельности
'''

In [29]:
# результат запроса
emp_without_ind = pd.read_sql_query(query_5_4, connection)
print(emp_without_ind)

   count
0   8419


<ipython-input-29-81662082ddce>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_without_ind = pd.read_sql_query(query_5_4, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [30]:
# текст запроса
query_5_5 = f'''
select
    e.name
from employers e --к таблице рвботодателей
--присоединяем таблицу со сферами деятельности
left join employers_industries ei on e.id = ei.employer_id --по id работодателя
group by 1
having count(ei.industry_id) = 4 --фильтруем по по количеству сфер
order by 1 --сортируем по алфавиту
offset 2 --обрезаем первые две строки
limit 1 --и ограничиваем вывод одной (третьей - искомой) строкой
'''

In [31]:
# результат запроса
emp_with_4_ind = pd.read_sql_query(query_5_5, connection)
print(emp_with_4_ind)

<ipython-input-31-64d1d795eab0>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_with_4_ind = pd.read_sql_query(query_5_5, connection)


   name
0  2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''
select
    count(e.employer_id)
from employers_industries e
--связываем таблицы работодателей и сфер деятельности
left join industries i on e.industry_id = i.id --по id сферы деятельности
--фильтруем по названию конкретной сферы
where i.name like '%Разработка программного обеспечения%'
'''

In [33]:
# результат запроса
software_dev = pd.read_sql_query(query_5_6, connection)
print(software_dev)

<ipython-input-33-c8644cc3112f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  software_dev = pd.read_sql_query(query_5_6, connection)


   count
0   3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8).

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API.

In [34]:
# код для получения списка городов-милионников

# Импортируем необходимые библиотеки
import requests
from bs4 import BeautifulSoup

# Задаём url страницы
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
# Выполняем GET-запрос, присваивая содержимое ответа переменной "response"
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')

# Находим первый объект под тегом 'table' (таблица) на странице
table = page.find_all('table')[0]

# Извлекаем строки таблицы кроме заголовка
rows = table.find_all('tr')[1:]

# Создаём пустой список
mill_city_list = []

# Проходим по всем строкам таблицы,
# извлекая текст только из второго столбца
# и добавляя его в список
for row in rows:
    mill_city_list.append(row.find_all('td')[1].text)

# преобразуем список в кортеж для sql-запроса
mill_city_list = tuple(mill_city_list)

In [35]:
# текст запроса
query_5_7 = f'''
select
    a.name,
    count(v.id) cnt
from vacancies v --к таблице вакансий
--присоединяем таблицы с регионами и работодателями
join areas a on a.id = v.area_id
join employers e on e.id = v.employer_id
--задаём фильтр по названию работодателя
where a.name in {mill_city_list} and e.name = 'Яндекс'
group by 1 --группируем по названиям городов
--объединяем с ещё одним запросом для вывода общего количества вакансий
union all
select
    'Total', --задаём значение первой ячейки
    count(v.id) --и считаем общее количество
from vacancies v
join areas a on a.id = v.area_id
join employers e on e.id = v.employer_id
--также фильтруем по названию конкретного работодателя
where a.name in {mill_city_list} and e.name = 'Яндекс'
order by 2 --сортируем всю получившуюся таблицу по количеству вакансий (по возрастанию)
'''

In [36]:
# результат запроса
yandex_mill_vac = pd.read_sql_query(query_5_7, connection)
print(yandex_mill_vac)

<ipython-input-36-2c4f6d656a04>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  yandex_mill_vac = pd.read_sql_query(query_5_7, connection)


               name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485


***

# Выводы по анализу работодателей

Топ-5 работодателей выглядит так:

1.	Яндекс		1933
2.	Ростелеком		491
3.	Тинькофф		444
4.	СБЕР			428
5.	Газпром нефть	331

Топ-5 по количеству работодателей в регионах:

1.	Москва		5,864 (5333 вакансий)
2.	Санкт-Петербург	2,217 (2851 вакансий)
3.	Минск			1,115 (2112 вакансий)
4.	Алматы		721 (1892 вакансий)
5.	Екатеринбург		609 (1698 вакансий)

Компании с самой широкой географией: Яндекс (публикуют вакансии в 181 регионах), Ростелеком (152 региона) и Спецремонт (116 регионов).

35% работодателей не указали сферу деятельности в своих вакансиях.

Топ компаний по количеству сфер деятельности опубликованных вакансий:

1.	Модуль		16
2.	Простые решения	11
3.	Энергия		10
4.	Прайд			10
5.	СБК			9

У компании Яндекс размещено 485 вакансий во всех городах-миллионерах, где лидерами по количеству вакансий ожидаемо являются Москва (54 вакансии) и Санкт-Петербург (42 вакансии). Далее идут Екатеринбург (39) и Нижний Новгород (36), а замыкает пятёрку Новосибирск с 35-ью вакансиями.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [37]:
# текст запроса

query_6_1 = f'''
select
    count(*) --считаем все строки
from vacancies --в таблице вакансий
where lower(name) like '%data%' or lower(name) like '%данн%' --содержащие необходимые сочетания
'''

In [38]:
# результат запроса
data_vac = pd.read_sql_query(query_6_1, connection)
print(data_vac)

<ipython-input-38-0af85e12a8fe>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_vac = pd.read_sql_query(query_6_1, connection)


   count
0   1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.


In [39]:
# текст запроса
query_6_2 = f'''
select
    count(name)
from vacancies
--применяем все необходимые фильтры по искомым сочетаниям слов
where((name ilike '%data scientist%'
  or name ilike '%data science%'
  or name ilike '%исследователь данных%'
  or (name like '%ML%' and name not ilike '%HTML%')
  or name ilike '%machine learning%'
  or name ilike '%машинн%обучен%')
  and (name ilike '%junior%' or experience ilike '%Нет опыта%' or employment ilike '%Стажировка%'))
'''

In [40]:
# результат запроса
ds_jun_vac = pd.read_sql_query(query_6_2, connection)
print(ds_jun_vac)

<ipython-input-40-5d6ceb6a6147>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_jun_vac = pd.read_sql_query(query_6_2, connection)


   count
0     51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
# текст запроса
query_6_3 = f'''
select
    count(name)
from vacancies
where((name ilike '%data scientist%'
  or name ilike '%data science%'
  or name ilike '%исследователь данных%'
  or (name like '%ML%' and name not ilike '%HTML%')
  or name ilike '%machine learning%'
  or name ilike '%машинн%обучен%')
  and (key_skills like '%SQL%' or key_skills like 'postgres'))
'''

In [42]:
# результат запроса
ds_sql_requirement = pd.read_sql_query(query_6_3, connection)
print(ds_sql_requirement)

<ipython-input-42-d9d213265953>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_sql_requirement = pd.read_sql_query(query_6_3, connection)


   count
0    201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
# текст запроса
query_6_4 = f'''
select
    count(name)
from vacancies
where((name ilike '%data scientist%'
  or name ilike '%data science%'
  or name ilike '%исследователь данных%'
  or (name like '%ML%' and name not ilike '%HTML%')
  or name ilike '%machine learning%'
  or name ilike '%машинн%обучен%')
  and (key_skills like '%Python%'))
'''

In [44]:
# результат запроса
ds_python_requirement = pd.read_sql_query(query_6_4, connection)
print(ds_python_requirement)

   count
0    351


<ipython-input-44-08cae1e73fb2>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_python_requirement = pd.read_sql_query(query_6_4, connection)


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
# текст запроса
query_6_5 = f'''
select
    --считаем количество слов в ячейках
    --по количеству пробелов CHR(9) между значениями
    --для этого от общей длины строки отнимаем длину строки без пробелов
    --добавляя ещё одно значение, т.к. после последнего пробела уже нет
    round(avg(length(key_skills) - length(replace(key_skills,CHR(9),''))+1),2)
from vacancies
where((name ilike '%data scientist%'
  or name ilike '%data science%'
  or name ilike '%исследователь данных%'
  or (name like '%ML%' and name not ilike '%HTML%')
  or name ilike '%machine learning%'
  or name ilike '%машинн%обучен%')
  and key_skills is not null)
'''

In [46]:
# результат запроса
key_skills_avg = pd.read_sql_query(query_6_5, connection)
print(key_skills_avg)

<ipython-input-46-7c6d7b0fdfb0>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  key_skills_avg = pd.read_sql_query(query_6_5, connection)


   round
0   6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*).

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [47]:
# текст запроса
query_6_6 = f'''
select
    experience "Опыт работы",
    --используя coalesce для перебора значений, считаем:
    --зп от + зп до, делённые на 2, если указаны оба значения
    --если одно из значений null
    --берём значение "от", если указано оно
    --или значение "до", если "от" - пустое
    round(avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to))) "Средняя з/п"
from vacancies
where((name ilike '%data scientist%'
  or name ilike '%data science%'
  or name ilike '%исследователь данных%'
  or (name like '%ML%' and name not ilike '%HTML%')
  or name ilike '%machine learning%'
  or name ilike '%машинн%обучен%')
  and (salary_from is not null or salary_to is not null)
 )
  group by 1
'''

In [48]:
# результат запроса
avg_salary = pd.read_sql_query(query_6_6, connection)
print(avg_salary)

<ipython-input-48-da33300a421f>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary = pd.read_sql_query(query_6_6, connection)


          Опыт работы  Средняя з/п
0           Нет опыта      74643.0
1  От 1 года до 3 лет     139675.0
2       От 3 до 6 лет     243115.0


***

# Выводы по предметному анализу
- На момент выгрузки датасета количество вакансий, имеющих то или иное отношение к работе с данными, составляет 3.5% от общего числа почти в 50.000, что довольно прилично, учитывая, что общее количество сфер деятельности, указанных в базе данных, составляет почти 300 шт.
Количество вакансий непосредственно в области Data Science составляет 480 шт, из которых 10% - вакансии для начинающих специалистов, что указывает на активный рост востребованности данного направления.
- Почти половина вакансий на позицию Data Scientist предполагает знание SQL и более 70% - Python, что подтверждает факт того, что этот язык программирования является самым популярным в области обработки данных.
- Заработные платы на данной позиции начинаются в среднем от 74.000 для новичков и растут по мере увеличения опыта работы в сфере – так при опыте от 1 года до 3-х в среднем имеют показатель около 140.000 и более 240.000 для опытных специалистов от 3-х до 6-ти лет.


# Общий вывод по проекту

**Итак, статистика нашей базы данных выглядит следующим образом:**
    
1. Вакансии - 49197
2. Работодатели - 23501
3. Регионы - 1362
4. Сферы деятельности - 294

**Топ-5 работодателей по количеству вакансий:**

1.  Яндекс      1933
2.  Ростелеком      491
3.  Тинькофф        444
4.  СБЕР            428
5.  Газпром нефть   331

**Топ-5 регионов по количеству вакансий:**
1. Москва
2. Санкт-Петербург
3. Минск
4. Новосибирск
5. Алматы

**Количество вакансий с удалённым графиком работы составляет 19%, что говорит о росте популярности данного типа занятости:**


In [49]:
query_7_1 = f'''
select
    count(*) --считаем общее количество строк
from vacancies --в таблице с вакансиями
where schedule = 'Удаленная работа' --используя фильтр по "удалёнке"
'''

dist_schedule = pd.read_sql_query(query_7_1, connection)
# Выводим результат, посчитав долю от общего количества вакансий
print(round(dist_schedule / 49197 * 100))

   count
0   19.0


<ipython-input-49-ee2e2fd1c71c>:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  dist_schedule = pd.read_sql_query(query_7_1, connection)


### Вакансии Data Science
Больше всего вакансий на позицию специалиста по Data Science ожидаемо размещено в Москве (217 шт.), далее следуют Санкт-Петербург (64 шт.) и Новосибирск (23 шт.) с Нижним Новгородом (20 шт.):

In [50]:
query_7_2 = f'''
select
    a.name,
    count(v.id)
from vacancies v
join areas a on a.id = v.area_id
where(v.name ilike '%data scientist%'
  or v.name ilike '%data science%'
  or v.name ilike '%исследователь данных%'
  or (v.name like '%ML%' and v.name not ilike '%HTML%')
  or v.name ilike '%machine learning%'
  or v.name ilike '%машинн%обучен%')
group by 1
order by 2 desc
'''

ds_area = pd.read_sql_query(query_7_2, connection)
print(ds_area)

<ipython-input-50-dd4ddaeb1394>:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_area = pd.read_sql_query(query_7_2, connection)


                         name  count
0                      Москва    217
1             Санкт-Петербург     64
2                 Новосибирск     23
3             Нижний Новгород     20
4                      Казань     16
5                      Алматы     16
6                       Минск      9
7                       Томск      8
8                Екатеринбург      7
9                   Краснодар      6
10             Ростов-на-Дону      6
11                    Армения      6
12                 Нур-Султан      6
13                     Самара      5
14                      Пермь      5
15                       Кипр      5
16                     Рязань      4
17                    Ташкент      4
18                  Иннополис      3
19                    Тбилиси      3
20                Владивосток      3
21                     Сербия      3
22                       Омск      3
23                        Уфа      2
24                    Израиль      2
25                     Турция      2
2

### Статистика вакансий по самым востребованным языкам программирования:

In [51]:
query_7_3 = f'''

--объединяем таблицы с фильтрами каждого интересующего языка программирования
--данные о самых популярных языках программирования взяты из статистики
--сайта "tproger" за 2024 год

select
    'Python' "Язык программирования",
    count(name) "Количество вакансий"
from vacancies
where key_skills ilike '%Python%'

union all
select
    'JAVA',
    count(name)
from vacancies
where key_skills ilike '%JAVA%' and (not key_skills ilike '%JavaScript%' or key_skills not ilike '%Java Script%')

union all
select
    'Java Script',
    count(name)
from vacancies
where key_skills ilike '%JavaScript%' or key_skills ilike '%Java Script%'

union all
select
    'C++',
    count(name)
from vacancies
where key_skills ilike '%C++%'

union all
select
    'C#',
    count(name)
from vacancies
where key_skills ilike '%C#%'

union all
select
    'PHP',
    count(name)
from vacancies
where key_skills ilike '%PHP%'

union all
select
    'Swift',
    count(name)
from vacancies
where key_skills ilike '%Swift%'
order by 2 desc
'''

language_stat = pd.read_sql_query(query_7_3, connection)
print(language_stat)

<ipython-input-51-a2628004040e>:57: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  language_stat = pd.read_sql_query(query_7_3, connection)


  Язык программирования  Количество вакансий
0                  JAVA                 5769
1           Java Script                 3718
2                Python                 3461
3                   PHP                 2052
4                    C#                 1521
5                   C++                 1162
6                 Swift                  457


1. JAVA	(5769 вакансий)
2. Java Script (3718 вакансий)
3. Python	(3461 вакансий)
4. PHP	(2052 вакансий)
5. C#	(1521 вакансий)
6. C++	(1162 вакансий)
7. Swift (457 вакансий)

### Самые популярные сферы деятельности:

In [52]:
query_7_4 = f'''
select
    i.name "Сфера деятельности",
    count(v.id) "Количество вакансий"
from vacancies v
join employers_industries ei on v.employer_id = ei.employer_id
join industries i on i.id = ei.industry_id
group by 1
order by 2 desc
limit 5
'''

top_industries = pd.read_sql_query(query_7_4, connection)
print(top_industries)

<ipython-input-52-af6a4dee8446>:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_industries = pd.read_sql_query(query_7_4, connection)


                                  Сфера деятельности  Количество вакансий
0                Разработка программного обеспечения                12499
1  Системная интеграция,  автоматизации технологи...                11034
2  Интернет-компания (поисковики, платежные систе...                 6413
3                                               Банк                 2742
4                                    Мобильная связь                 1585


1. Разработка программного обеспечения (12499 вакансий)
2. Системная интеграция,  автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг (11034 вакансий)
3. Интернет-компания (поисковики, платежные системы, соц.сети, информационно-познавательные и развлекательные ресурсы, продвижение сайтов и прочее) (6413 вакансий)
4. Банк (2742 вакансий)
5. Мобильная связь (1585 вакансий)

### Самые прибильные сферы (по средней з/п):

In [53]:
query_7_5 = f'''
select
    i.name "Сфера деятельности",
    --вычисляем среднюю ЗП аналогично запросу 6_6
    round(avg(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to))) "Средняя з/п"
from vacancies v
join employers_industries ei on v.employer_id = ei.employer_id
join industries i on i.id = ei.industry_id
where v.salary_from is not null or v.salary_to is not null
group by 1
order by 2 desc
limit 5
'''

top_salaries = pd.read_sql_query(query_7_5, connection)
print(top_salaries)

                          Сфера деятельности  Средняя з/п
0  Агрохимия (продвижение, оптовая торговля)     381429.0
1                   Агрохимия (производство)     339875.0
2                             Игорный бизнес     191905.0
3          Корма для животных (производство)     168300.0
4                                 Зоомагазин     167613.0


<ipython-input-53-c68db1b36827>:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_salaries = pd.read_sql_query(query_7_5, connection)


1. Агрохимия (продвижение, оптовая торговля) - 381429 руб/мес
2. Агрохимия (производство) - 339875.0 руб/мес
3. Игорный бизнес - 191905.0 руб/мес
4. Корма для животных (производство) - 168300.0 руб/мес
5. Зоомагазин - 167613.0 руб/мес

In [54]:
connection.close()